### PR TITLE
Fix `NumPyEigensolver` with `SparsePauliOp`

### LINKED ISSUE #9091: `NumPyEigensolver` does not support all `BaseOperator` instances.
### Environment

- **Qiskit Terra version**: 3ce1737b
- **Python version**: 3.10.6
- **Operating system**: macOS 13.0

### What is happening?

`NumPyEigensolver` and by extension `NumPyMinimumEigensolver` do not support all `BaseOperator` instances, in particular `SparsePauliOp`. This is because it requires the `data` attribute, which `Operator` exposes but e.g. `SparsePauliOp` does not. 

### How can we reproduce the issue?

```python
from qiskit.algorithms.eigensolvers import NumPyEigensolver
from qiskit.quantum_info import SparsePauliOp

op = SparsePauliOp("Z")
npe = NumPyEigensolver()
result = npe.compute_eigenvalues(op)
```

Output:
```
Traceback (most recent call last):
  File "/Users/declanmillar/Projects/qiskit/qiskit-terra/test.py", line 10, in <module>
    result = npe.compute_eigenvalues(op)
  File "/Users/declanmillar/Projects/qiskit/qiskit-terra/qiskit/algorithms/eigensolvers/numpy_eigensolver.py", line 248, in compute_eigenvalues
    self._solve(operator)
  File "/Users/declanmillar/Projects/qiskit/qiskit-terra/qiskit/algorithms/eigensolvers/numpy_eigensolver.py", line 144, in _solve
    if operator.data.all() == operator.data.conj().T.all():
AttributeError: 'SparsePauliOp' object has no attribute 'data'
```

### What should happen?

The code should compute the eigenvalues without error, similarly to:

```python
from qiskit.algorithms.eigensolvers import NumPyEigensolver
from qiskit.quantum_info import SparsePauliOp
from qiskit.opflow import PauliSumOp

op = SparsePauliOp("Z")
op = PauliSumOp(op)
npe = NumPyEigensolver()
result = npe.compute_eigenvalues(op)
```

or:

```python
from qiskit.algorithms.eigensolvers import NumPyEigensolver
from qiskit.quantum_info import SparsePauliOp, Operator

op = SparsePauliOp("Z")
op = Operator(op.to_matrix())
npe = NumPyEigensolver()
result = npe.compute_eigenvalues(op)
```

### Any suggestions?

Generalize the computation inside `NumPyEigensolver` to avoid the `data` attribute, perhaps via `to_matrix()` or, if possible, a sparse implementation that works for `BaseOperator` instances.

### PR DESCRIPTION
**Pull request by declanmillar**
### Summary

Fixes an issue where `NumPyEigensolver` and by extension `NumPyMinimumEigensolver` did not support all `BaseOperator` instances, in particular `SparsePauliOp`. This was because it required the data attribute, which `Operator` exposes but e.g. `SparsePauliOp` does not. Closes #9091.

### Details and comments

For all `BaseOperator` child classes, other than `Operator`, `to_matrix` is used to get the `NumPy` matrix for the operator and compute the eigenvalues and eigenstates. If an `Operator` instance, we still use `data`.

Additionally this PR:
* Refactors `NumPyEigensolver` to be stateless w.r.t. results. I.e. the result is no longer stored.
* Adds unit tests to check `NumPyEigensolver` with `SparsePauliOp`s.
* Adds units tests to check `NumPyMinimumEigensolver` with `SparsePauliOp`s and `Operator`s.
* Uses sparse matrix computation for all operators that support it, otherwise try dense computation. If this fails and error is raised.

### CODE REVIEW

**Comment by Cryoris** [comment]
It might be good to try using the sparse eigensolver (and the sparse matrix multiplication in the eval_aux_ops) when possible. Right now this is implemented for the `PauliSumOp`, but could definitely also be done for the `SparsePauliOp` (or maybe even all non-`Operator` operators?) by using `to_matrix(sparse=True)` 🙂

File: qiskit/algorithms/eigensolvers/numpy_eigensolver.py [hunk 1]
```diff
@@ -257,36 +237,35 @@ def compute_eigenvalues(
...
-            self._ret.aux_operators_evaluated = aux_ops
+            result.aux_operators_evaluated = aux_ops
 
             self._k = k_orig
 
-        # evaluate ground state after filtering (in case a filter is set)
-        self._get_ground_state_energy(operator)
-        if self._ret.eigenstates is not None:
-            self._ret.eigenstates = [Statevector(vec) for vec in self._ret.eigenstates]
+        if result.eigenstates is not None:
```
Comment (Cryoris) [inline qiskit/algorithms/eigensolvers/numpy_eigensolver.py:263]: It seems that this is never None, so would it make sense to immediately cast to a `Statevector` above?

**Review by Cryoris — COMMENTED** [review]
LGTM, some minor comments on structure below !

File: qiskit/algorithms/eigensolvers/numpy_eigensolver.py [hunk 2]
```diff
@@ -194,21 +174,17 @@ def _eval_aux_operators(
...
-                            * operator.coeff
-                        )
-            else:
+            if isinstance(operator, Operator):
                 value = Statevector(wavefn).expectation_value(operator)
+            else:
+                if isinstance(operator, PauliSumOp):
+                    if operator.coeff != 0:
+                        op_matrix = operator.to_spmatrix()
+                        value = op_matrix.dot(wavefn).dot(np.conj(wavefn))
```
[the remark this answers is not in the document]
Comment (declanmillar) [reply qiskit/algorithms/eigensolvers/numpy_eigensolver.py:183]: Done. It does not require an extra `None` check though for cases where `operator.coeff == 0`, which can be removed when `PauliSumOp` is retired.

File: qiskit/algorithms/eigensolvers/numpy_eigensolver.py [hunk 3]
```diff
@@ -110,68 +108,50 @@ def _check_set_k(self, operator: BaseOperator | PauliSumOp) -> None:
                 self._k = self._in_k
 
     def _solve(self, operator: BaseOperator | PauliSumOp) -> None:
-        if isinstance(operator, PauliSumOp):
-            sp_mat = operator.to_spmatrix()
+        if isinstance(operator, Operator):
```
Comment (woodsp-ibm) [inline qiskit/algorithms/eigensolvers/numpy_eigensolver.py:111]: This is a more general comment. `BaseOperator | PauliSumOp` was chosen to be the same as that which Estimator supports. However from what I can see BaseOperator provides no standard way to access the data comprising the operator as needed here. An Operator has data, and Pauli types seems to have to_matrix(). ScalarOp is also a BaseOperator and while it has to_matrix does not have a sparse attribute hence would fail if used here. Looking at Estimator when it does an expectation, using StateVector, it may end up creating an Operator, depending on type, where its constructor inspects the object being passed for various aspects to build it out - I think ScalarOp would pass through this. I guess I am wondering more about the if its not an Operator and then its not PauliSumOp we are in the realm of all other BaseOperators that are not the Operator subclass thereof. The to_matrix(sparse=True) looks like for the Pauli based ones like PauliTable, SparsePauliOp that this is implemented in these - but again this is a specific set of these. This logic certainly now captures the Pauli types like SparsePauliOp of BaseOperator, which it did not before. Just wondering whether we can/should do better - maybe at least wrap the to_matrix(sparse=Try) in a try/except and raise perhaps an error if that can't be called that the operator type is not supported?

Comment (declanmillar) [reply qiskit/algorithms/eigensolvers/numpy_eigensolver.py:111]: It looks like a try/except is the best option, as there are several operator classes that do not subclass `BasePauli` but do allow `sparse=True`. It will also give a more appropriate error when a parameterized operator is passed in. From what I can see all `BaseOperator` classes, other than `Operator`, do expose a `to_matrix` method. Would it be worth adding this method to `Operator` and just returning `data`? `Clifford` and `CNOTDihedral` already effectively do this via `to_operator`. Could we perhaps also make `to_matrix` a required method of `BaseOperator`.

Comment (woodsp-ibm) [reply qiskit/algorithms/eigensolvers/numpy_eigensolver.py:111]: Since primitives - especially the remote ones, I would think, need access to the data to be able to do the computation the fact that BaseOperator is the designated type but provides no defined mechanisms, indeed seems problematic. I am tagging @ikkoham and @t-imamichi to bring them into the conversation more generally. The reference Estimator relies on StateVector being able to compute an estimation of the operator (it and the operators being part of quantum_info) - for other primitives, that may use an aspect of it, like here, it can be different.

Here specifically, in the near term, I think a try/except to be able to raise a more meaningful error. Longer term it seems to me that whatever the type we indicate should have suitable method(s) for the purposes we have/envision. Whether BaseOperator (or something else) is updated or perhaps new mixin/Protocol(s) added seems like it should be the subject of a separate issue/PR to discuss and hopefully address these wider needs and revisit this try/catch later as warrants.

**Review by ElePT — COMMENTED** [review]
The code LGTM. I thought that maybe it would be possible to wrap all inputs in a `SparsePauliOp` not to have to deal with each case separately, but I could not find a better way to do it. I have followed a simular principle on this PR for VQD #9245.

File: test/python/algorithms/eigensolvers/test_numpy_eigensolver.py [hunk 4]
```diff
@@ -20,26 +20,29 @@
...
-
-H2_PAULI = PauliSumOp.from_list(
-    [
-        ("II", -1.052373245772859),
-        ("ZI", 0.39793742484318045),
-        ("IZ", -0.39793742484318045),
-        ("ZZ", -0.01128010425623538),
-        ("XX", 0.18093119978423156),
-    ]
+from qiskit.quantum_info import Operator, SparsePauliOp
```
[the remark this answers is not in the document]
Comment (declanmillar) [reply test/python/algorithms/eigensolvers/test_numpy_eigensolver.py:23]: There are additional outliers that cause issues. For example `PauliList` and `PauliTable`, which also subclass `BaseOperator` but will cause an error in `NumPyEigensolver`. We could accommodate these, but they would require individual logic. This is easy enough, but seems like a lot of extra code for functionality that no one is likely to need.

[the remark this answers is not in the document]
Comment (ElePT) [reply test/python/algorithms/eigensolvers/test_numpy_eigensolver.py:23]: No, I don't think it makes sense to add logic for edge cases... but we could check that the error raised is informative enough and these cases are accounted for (even if through raising an error).

**Review by ElePT — COMMENTED** [review]
LGTM! @Cryoris, if the CI doesn't break, I would say it's mergeable 😇

**Review by Cryoris — APPROVED** [review]
